### PR TITLE
Add traverseBia rules for containers types

### PR DESCRIPTION
**Pull request by treeowl**
* Add `RULES` to make `traverseBia` more efficient for
  `Data.Map`, `Data.IntMap`, `Data.Sequence`, and `Data.Tree`.

* Monkey around a bit to try to get the rules to fire more
  reliably. I *think* it worked, but this all feels like some
  sort of dark art.

### CODE REVIEW

File: src/Data/Biapplicative.hs [hunk 1]
```diff
@@ -259,6 +277,119 @@ traverseBiaConst _f (Const x) = bipure (Const x) (Const x)
...
+traverseBiaIntMap f = go
+  where
+    go IM.Nil = bipure IM.Nil IM.Nil
+    go (IM.Tip k v) = bimap (IM.Tip k) (IM.Tip k) (f v)
+    go (IM.Bin p m l r)
+      | m < 0     = biliftA2 (flip (IM.Bin p m)) (flip (IM.Bin p m)) (go r) (go l)
+      | otherwise = biliftA2 (IM.Bin p m) (IM.Bin p m) (go l) (go r)
+
+{-# INLINE traverseBiaTree #-}
+traverseBiaTree :: Biapplicative p => (a -> p b c) -> Tree.Tree a -> p (Tree.Tree b) (Tree.Tree c)
```
Comment (RyanGlScott) [inline src/Data/Biapplicative.hs:305]: Should `traverseBiaTree` exist outside of this `#if MIN_VERSION_containers(0,5,8) ... #endif` block? The `traverseBia/Tree` rewrite rule is not declared within CPP, so it appears to be mismatched.

**Review by RyanGlScott — COMMENTED** [review]
LGTM, aside from one small suggestion.

File: src/Data/Biapplicative.hs (hunk 1 above)
Comment (treeowl) [reply src/Data/Biapplicative.hs:305]: Oops...

**Review by RyanGlScott — APPROVED** [review]

**Comment by treeowl** [comment]
Thanks!

**Pull request merged**